rllib/agents/dqn/r2d2_tf_policy.py

```diff
@@ -0,0 +1,346 @@
+    TrainerConfigDict
+
+tf1, tf, tfv = try_import_tf()
+
+
+def build_r2d2_model(policy: Policy, obs_space: gym.spaces.Space,
+                     action_space: gym.spaces.Space, config: TrainerConfigDict
+                     ) -> Tuple[ModelV2, ActionDistribution]:
+    """Build q_model and target_q_model for DQN
+
+    Args:
+        policy (Policy): The policy, which will use the model for optimization.
+        obs_space (gym.spaces.Space): The policy's observation space.
+        action_space (gym.spaces.Space): The policy's action space.
+        config (TrainerConfigDict):
+
+    Returns:
+        q_model
+            Note: The target q model will not be returned, just assigned to
+            `policy.target_q_model`.
+    """
+
+    # Create the policy's models.
+    model = build_q_model(policy, obs_space, action_space, config)
+
+    # Assert correct model type.
+    assert model.get_initial_state() != [], \
+        "R2D2 requires its model to be a recurrent one! Try using " \
+        "`model.use_lstm` or `model.use_attention` in your config " \
+        "to auto-wrap your model with an LSTM- or attention net."
+
+    return model
+
+
+def r2d2_loss(policy: Policy, model, _,
+              train_batch: SampleBatch) -> TensorType:
+    """Constructs the loss for R2D2TFPolicy.
+
+    Args:
+        policy (Policy): The Policy to calculate the loss for.
+        model (ModelV2): The Model to calculate the loss for.
+        train_batch (SampleBatch): The training data.
+
+    Returns:
+        TensorType: A single loss tensor.
+    """
+    config = policy.config
+
+    # Construct internal state inputs.
+    i = 0
+    state_batches = []
+    while "state_in_{}".format(i) in train_batch:
+        state_batches.append(train_batch["state_in_{}".format(i)])
+        i += 1
+    assert state_batches
+
+    # Q-network evaluation (at t).
+    q, _, _, _ = compute_q_values(
+        policy,
+        policy.q_model,
+        train_batch,
+        state_batches=state_batches,
+        seq_lens=train_batch.get("seq_lens"),
+        explore=False,
+        is_training=True)
+
+    # Target Q-network evaluation (at t+1).
+    q_target, _, _, _ = compute_q_values(
+        policy,
+        policy.target_q_model,
+        train_batch,
+        state_batches=state_batches,
+        seq_lens=train_batch.get("seq_lens"),
+        explore=False,
+        is_training=True)
+
+    if not hasattr(policy, "target_q_func_vars"):
+        policy.target_q_func_vars = policy.target_q_model.variables()
+
+    actions = tf.cast(train_batch[SampleBatch.ACTIONS], tf.int64)
+    dones = tf.cast(train_batch[SampleBatch.DONES], tf.float32)
+    rewards = train_batch[SampleBatch.REWARDS]
+    weights = tf.cast(train_batch[PRIO_WEIGHTS], tf.float32)
+
+    B = tf.shape(state_batches[0])[0]
+    T = tf.shape(q)[0] // B
+
+    # Q scores for actions which we know were selected in the given state.
+    one_hot_selection = tf.one_hot(actions, policy.action_space.n)
```

Comment:
Support for Continuous? i.e. did R2d2 work beyond atari?

Reply:
Nope, it's classic Q-learning. We could - in the future - try an "SAC/DDPG R2D2" and also a "MARWIL R2D2". We'll see.

Reply:
That can be an entire paper LOL